### PR TITLE
Fixes 589

### PR DESCRIPTION
**Pull request by delvedor**
As titled, see #589 for details.

### CODE REVIEW

**Comment by jsumners** [comment]
Is this a breaking change?

File: fastify.js [hunk 1]
```diff
@@ -436,6 +436,9 @@ function build (options) {
       preHandler.push.apply(preHandler, opts.preHandler || _fastify._hooks.preHandler)
```
Comment (StarpTech) [inline fastify.js]: Correct If I'm wrong but why we add the prehandler twice. Here and after the loop.

Comment (allevo) [reply fastify.js]: I've edited the test in order to check is the `beforeHandler` is executed twice

**Comment by nwoltman** [comment]
Aren't all hooks supposed to be bound to the `fastify` instance? I thought that's what [this part of the documentation](https://github.com/fastify/fastify/blob/master/docs/Hooks.md#scope) meant. The `context` object is already available on the `reply` object.

**Comment by jsumners** [comment]
@nwoltman that's what this does. Notice in the test that he decorates the Fastify instance with a property `foo`. He then accesses that property in the `beforeHandler` via `this`. In other words, `this` in the `beforeHandler` is `fastify`.

**Comment by nwoltman** [comment]
@jsumners Sorry for not being more clear, I know that's what these changes do. I was just really confused why [the other hooks](https://github.com/fastify/fastify/blob/5ac50d37550ee0e5218e47d19fa724299da0ed71/fastify.js#L442-L445) are getting bound to `context` instead of `_fastify` (the second argument to `runHooks`/[`fast-iterator`](https://github.com/delvedor/fast-iterator) is supposed to be the context that the functions will run in).

After some investigation, I found that hooks added by [`fastify.addHooks()`](https://github.com/fastify/fastify/blob/5ac50d37550ee0e5218e47d19fa724299da0ed71/fastify.js#L556-L563) get bound to the Fastify instance in that function (which means that when they get passed to `fast-iterator` they get [bound again](https://github.com/delvedor/fast-iterator/blob/master/index.js#L7), which doesn't do anything since they're already bound).

It seems like binding in `fastify.addHooks()` and the changes in this PR would be unnecessary if `_fastify` was passed as the second argument to `runHooks` instead of `context`. Either that or passing `context` to `runHooks` is unnecessary.

File: fastify.js (hunk 1 above)
Comment (StarpTech) [reply fastify.js]: My mistake it's different.

**Comment by jsumners** [comment]
Sounds like a vestige that didn't get refactored properly.

**Comment by mcollina** [comment]
👍  that all that part need a big refactoring. We have changed shapes of things very often in the hooks area, and the logic is more complicated than it needs to be.

I'm going to land this, feel free to send a PR for refactoring things @nwoltman!

**Pull request merged**